### DISCUSSION TITLE
fixed width tables do not get parsed in docx

### DISCUSSION THREAD

**Discussion by luerhard** [ickc/pantable#64] [Q&A] [answered]
Hey there, i am currently trying to use pantable to export tables into a couple of different formats. 

When I export this table
```
\```table
---
alignment: LLL
markdown: true
width: [0.2, 0.5, 0.3]
---
1,2,3
hello,my,name
\```
```

to latex, I get the following latex code:

```latex
\begin{longtable}[]{@{}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.2000}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.5000}}
  >{\raggedright\arraybackslash}p{(\columnwidth - 4\tabcolsep) * \real{0.3000}}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
1
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
2
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
3
\end{minipage} \\
\midrule
\endhead
hello & my & name \\
\bottomrule
\end{longtable}
```

When I try to use pandoc to convert this to a docx file, the table does not get parsed (and the header is cut off).

![image](https://user-images.githubusercontent.com/24412608/146568717-bd034b94-efa2-4c1e-8591-5300a7eead49.png)

This definitely has to do with the `width: [0.2, 0.5, 0.3]`. Is this an issue with pantable or with pandoc?

**Comment by ickc** [comment, marked as answer]
I'm pretty sure it is not a bug of pantable.

One of the problem in trying to reproduce your result is that your example above is not self-contained. i.e. you do not have the command you use to generate the docx.

I also assume your `\` in the beginning is not what you meant (you may be trying to escape that in showing it as code in markdown here, which you can do will either using another delimiter such as `~~~` or even indent the whole block.)

In `test.md`:

~~~markdown
```table
---
alignment: LLL
markdown: true
width: [0.2, 0.5, 0.3]
---
1,2,3
hello,my,name
```
~~~

```bash
pandoc -F pantable test.md -o test.docx
```

is a normal looking table to me.

Another reason to know that this is not a problem of pantable is that your LaTeX output is correct—pantable only convert the table code-block to pandoc AST, and is output format agnostic.

Edit: also the versions (and sometimes platform, or even the software you use to open the docx) can be useful in reproducing. e.g. mine is pantable 0.14.2, pandoc 2.16.2, MS Word for Mac version 16.56

**Reply by ickc** [reply]
A reply from @luerhard was lost in GitHub, perhaps because coincidentally it was sent the moment this was converting to a Discussion. Copied below:

As a short follow-up: I cannot get the above Latex table created with pdflatex at all. The closest I can get is by doing:

\begin{longtable}[]{@{}
{\raggedright\arraybackslash}p{\dimexpr 0.2\columnwidth - 4\tabcolsep}
{\raggedright\arraybackslash}p{\dimexpr 0.5\columnwidth - 4\tabcolsep}
{\raggedright\arraybackslash}p{\dimexpr 0.3\columnwidth - 4\tabcolsep}@{}}
\toprule
\begin{minipage}[b]{\linewidth}\raggedright
       1
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
       2
\end{minipage} & \begin{minipage}[b]{\linewidth}\raggedright
       3
\end{minipage} \\
\midrule
\endhead
hello & my & name \\
\bottomrule
\end{longtable}

**Reply by ickc** [reply]
Again, I cannot reproduce the problem,

```bash
pandoc -F pantable test.md -o test.pdf
```

was fine for me.

This is also fine:

```bash
pandoc -F pantable test.md -s -o test.tex
latexmk -pdf test.tex
```

You may have missed `-s`.

**Comment by luerhard** [comment]
Yes, you are right. I was overwhelmed yesterday was a bit quick in opening an issue on pantable.

When I was testing pantable, I missed `\usepackage{calc}` in my latex minimal examples.
This is apparently loaded behind the scenes, when one uses the `todonotes` package, which is why the latex would parse with my regular templates, but not in minimal latex examples. Sorry for that.

The full list of package dependencies in Latex, I came up with, then is:
```latex
\usepackage{array}
\usepackage{booktabs}
\usepackage{calc}
\usepackage{longtable}
```


The minimal self-contained example for my problem then is

File `main.md`
~~~
```table
---
width: [0.2, 0.5, 0.3]
---
1,2,3
hello,my,name
```
~~~

does not produce a table in docx when using:

Command:
`pandoc -F pantable -o out.tex main.tex && pandoc -o out.docx main.tex`

But after spending a couple more hours with it yesterday, I understood that this is rather a problem with the Latex-Reader of pandoc than with pantable. So sorry for all the hectic and confusion.

In my usecase, I want to convert Markdown to latex, then run additional filters and stuff on it to just finally convert it from Latex to pdf and docx respectively. So that is why I have this very unusual/strange command (if you are wondering).

And thanks for the `~`tip -- that works great !

My Software versions:

Kubuntu 21.10, with pandoc 2.16.1, opening the docx files with LibreOffice.
It is the same on Manjaro Linux, with pandoc 2.14 (LibreOffice as well).

**Reply by ickc** [reply]
My advice would be use markdown as your source and all your conversion from there.

The reason is that the pandoc model is n readers to AST to m writers. So when you are authoring in pandoc, you want something as close to the AST as possible (or as close to losslessness.)

Eg pandoc markdown (and its AST) allows raw TeX, so you can use either LaTeX directly, or even use `{=tex}...` RawBlock syntax to enforce tex only (which will only shows up in TeX output. Targeting multiple format means writing multiple RawBlock for each output format. Docx would be difficult if not impossible.

Also all filters acts on the AST, so it needs not be written into a certain intermediate format. Chain them up as much as possible in one command. If you have to break it down for some reason, you can write to native/json and the next command read from that.

Hope that helps!

**Reply by luerhard** [reply]
Yes, this helped a lot -- thank you !

I rethought my pipeline over this, and indeed staying close to the AST made things so much easier. You revolutionized my pandoc usage and got me to undo a lot of stupid stuff that grew over time in my pipeline. 

I am quite new to pandoc and maybe I bit off more than I could chew for starters... Thanks again !
